### PR TITLE
Query captions/subtitles [e.g. video transcripts] from xklb-metadata.db

### PR DESCRIPTION
**Pull request by deldesir**
**🚀 Pull Request Overview:**
This is an alternative PR to #152 designed for A/B testing. It allows searching querying xklb-metadata.db. The ~video titles~ ids returned as results are used to enhance Calibre-Web's simple search.

cf. https://github.com/chapmanjacobd/library/blob/main/xklb/mediadb/search.py

**📋 Checklist:**
- [ ] Tested the changes on Ubuntu 24.04 with https://www.youtube.com/watch?v=V8Bn0UPpGCw using xklb v2.9.036
```
root@box:~# lb search /library/calibre-web/xklb-metadata.db phytoplankton --to-json
8 captions
Is this the most valuable thing in the ocean? - David Biello - /library/calibre-web/TED-Ed/Is this the most valuable thing in the ocean_ - David Biello (9)/Is this the most valuable thing in the oce - TED-Ed.mp4
    1:21 we need to look at some of the ocean’s most ubiquitous organisms:
         phytoplankton.
    1:41 phytoplankton are everywhere.. and phytoplankton blooms are large
         enough to be seen from space.. These phytoplankton then become
         food for countless microscopic grazers,
    2:11 But when phytoplankton die, their bodies can sink far below the
         surface,
    2:27 whales consume huge amounts of these phytoplankton predators.
    3:16 Phytoplankton’s massive photosynthetic blooms. phytoplankton
         helps sequester huge amounts of carbon—. According to one report,
         phytoplankton are estimated to capture four times. every 1%
         increase in phytoplankton population
    4:02 researchers are working to increase phytoplankton populations
    4:17 than supporting the ocean’s natural phytoplankton farmers.
```
image

**📌 Testing scenario:**
- Download a single video or playlist
- Use the search field to search for some word or phrase present in the subtitles. You might need to watch a video to get some terms ;-)

_Related to #152_ 

cc @EMG70

### CODE REVIEW

**Comment by holta** [comment]
Related:

- PR #152 
- #197

**Comment by holta** [comment]
Should this PR be tested more broadly while comparing it to PR #152 now, as suggested at... ?

- https://github.com/iiab/calibre-web/pull/152#issuecomment-2323057281

**Comment by deldesir** [comment]
It's still a draft. Trying to emulate xklb's internals here but my tests aren't conclusive yet due to a misunderstanding of how fts works. I am refactoring this completely to mirror (steal) xklb's code, then will remove the features we don't want for simplicity sake. Worth it as I am learning a lot by doing this.

**Comment by deldesir** [comment]
Ready for testing.

**Comment by holta** [comment]
> Ready for testing.

Thanks!

Can you summarize the differences from PR #152 ?

> This is an alternative PR to #152 designed for A/B testing. It allows searching querying xklb-metadata.db

Does this PR #244 search both metadata.db and xklb-metadata.db for completeness?

**Comment by deldesir** [comment]
> > Ready for testing.
> 
> Thanks!
> 
> Can you summarize the differences from PR #152 ?
> 
> > This is an alternative PR to #152 designed for A/B testing. It allows searching querying xklb-metadata.db
> 
> Does this PR #244 search both metadata.db and xklb-metadata.db for completeness?

No, as explained [in #197 ](https://github.com/iiab/calibre-web/issues/197#issuecomment-2324772178), this PR only searches xklb-metadata.db. Completeness is assured by calibre-web internal search which merges its own results with this PR's ones.

**Comment by holta** [comment]
> > Can you summarize the differences from PR #152 ?
> 
> This PR doesn't run any command in the background to get its results output to STDOUT, #152 runs a command.

@deldesir explain in a bit more detail please, for everyone?

How/why is this approach (PR #244) likely preferred over PR #152 ?

**Comment by holta** [comment]
@EMG70 would you have time to test this PR?

(And compare it to your impressions of PR #152 !)

**Comment by EMG70** [comment]
I will test later on Today.

**Comment by EMG70** [comment]
Please help if my testing is OK:
SUDO IIAB-DIAGNOSTICS - https://paste.centos.org/view/3c0d2ebe

I have installed a new VM and ran PR#244 ,but it does not return search results when a phrase or word is searched in GUI.
It however returns results via lb-wrapper search in CLI
I have tried this https://github.com/iiab/calibre-web/pull/152#issuecomment-2323677483 in case lb-wrapper is not updated as happened in my PR#152 test , but no luck.I may be testing wrongly ,please help.

**Comment by deldesir** [comment]
Did you do?

1- `cd /usr/local/calibre-web-py3`
2- `git pull https://github.com/deldesir/calibre-web deldesir-search-captions --no-rebase --no-edit`
3- `systemctl restart calibre-web`

No need to move lb-wrapper after this.

**Comment by EMG70** [comment]
> Did you do?
> 
> 1- `cd /usr/local/calibre-web-py3` 2- `git pull https://github.com/deldesir/calibre-web deldesir-search-captions --no-rebase --no-edit` 3- `systemctl restart calibre-web`
> 
> No need to [copy] lb-wrapper after this.

I had followed latest PR test instructions ,I should have done it afresh after it asked for email .That's where i probably missed something.

**Comment by EMG70** [comment]
Now tested and working as expected.

Screenshot from 2024-09-03 21-07-50
<pre>root@box:/# lb search /library/calibre-web/xklb-metadata.db mechanisms --to-json
1 captions
Button Mushroom Production in Zimbabwe - /library/calibre-web/Agribusiness With Wadzanai Manyore/Button Mushroom Production in Zimbabwe (1)/Button Mushroom Production in Zimbabwe - Agribusiness With Wadzanai Manyore.mp4
   17:19 you need to be able to use those you need to be able to use those mechanisms so uh working with them has. mechanisms so uh working with them has mechanisms so uh working
         with them has been really nice because I think our

root@box:/# 

</pre>

**Comment by holta** [comment]
@codewiz

@deldesir asks if this should perhaps be refactored as a class, e.g. to "ensure safety accessing its functions designed for module scope" ?

**Comment by deldesir** [comment]
> @codewiz
> 
> @deldesir asks if this should perhaps be refactored as a class, e.g. to "ensure safety accessing its functions designed for module scope" ?

Already structured as a class in https://github.com/iiab/calibre-web/pull/244/commits/5e273e4e62cd73e91b53664503c9b219fcbd7a6c

**Comment by codewiz** [comment]
A few questions:

- How many records can xklb-metadata.db contain on a large server?
- Does SQLite DB have an index on captions.text?
- Does the index work also for queries using LIKE, or will it end up doing a full scan?
- Should there be a "LIMIT 1000" to avoid OOMing the Python process in case a user searches for a common word, such as "%the%" or even "%a"%"?
- Would it make sense to reuse a long-lived CaptionSearcher instance to avoid the cost of reopening the DB on every query?

**Comment by avni** [comment]
TLDR: `PR #244` worked well! `PR #152` applied to a new #wrench-installation did not work.

====
I tested on 3 calibre-web instances with the video, [Alternating Series, Types of Convergence, and the Ratio Test](https://www.youtube.com/watch?v=-Md4CEW0-zM) , searching for the term "p-series". These are all new installs, of the [wrench installation](https://github.com/iiab/calibre-web/wiki#wrench-installation). Results: 

1. Neither `PR #244` or `PR #152` applied - search returned 0 results. ✅ 
2. `PR #244` applied - search returned video, quickly ✅ 
3. `PR #152` applied - search returned 0 results, it should return the video. 🤔. I'll post diagnostics in https://github.com/iiab/calibre-web/pull/152

**Comment by avni** [comment]
> `PR #152` applied to a new #wrench-installation did not work.

Updating lb-wrapper as described here fixed it: https://github.com/iiab/calibre-web/pull/152#issuecomment-2323677483. 

I'm curious though why `PR #244` uses the older version of lb-wrapper and the search still works. 🤔

**Comment by holta** [comment]
@deldesir please answer the [first 4 questions](https://github.com/iiab/calibre-web/pull/244#issuecomment-2327807092) from @codewiz above, as best you can 🙏 and we'll help asking the way ✅

Regarding @codewiz's Question 5. here...

> Would it make sense to reuse a long-lived CaptionSearcher instance to avoid the cost of reopening the DB on every query?

I'd highly doubt it.  At the moment anyway.

But if teachers ask all 32 kids in an Internet-in-a-Box classroom to search at the same time in future years... _then yes we might need to revisit our assumptions!_ 💫

**Comment by deldesir** [comment]
> A few questions:
> 
> * How many records can xklb-metadata.db contain on a large server?

Millions of records, I presume. SQLite databases are limited by file size, and technically the maximum size is 140 terabytes. I don't think we'll ever reach this (https://www.sqlite.org/limits.html)

> * Does SQLite DB have an index on captions.text?

It seems like there isn't. Doing `SELECT name, tbl_name, sql FROM sqlite_master WHERE type = 'index' AND tbl_name = 'captions';` outputs:

```
name,tbl_name,sql
idx_captions_media_id,captions,"CREATE INDEX [idx_captions_media_id]
    ON [captions] ([media_id])"
idx_captions_time,captions,"CREATE INDEX [idx_captions_time]
    ON [captions] ([time])"
```

> * Does the index work also for queries using LIKE, or will it end up doing a full scan?
It does a full scan

> * Should there be a "LIMIT 1000" to avoid OOMing the Python process in case a user searches for a common word, such as "%the%" or even "%a"%"?

I think so. I'll do it.

> * Would it make sense to reuse a long-lived CaptionSearcher instance to avoid the cost of reopening the DB on every query?

This makes sense, we already are dealing with OOM issues with the web server for large size videos.

File: cps/lb_search.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        query = f"""
+            SELECT media.path, media.title, captions.text, captions.time
+            FROM captions
+            JOIN media ON media.id = captions.media_id
+            WHERE captions.text LIKE ?
+            ORDER BY media.path, captions.time
+        """
+        return query, (f"%{term}%",)
+
+    def get_search_terms(self, term):
```
Comment (avni) [inline cps/lb_search.py:49]: Consider renaming this to get_captions_search_results or something similar to make clear this is searching captions.

File: cps/db.py [hunk 2]
```diff
@@ -973,7 +974,19 @@ def get_cc_columns(self, config, filter_config_custom_read=False):
     def get_search_results(self, term, config, offset=None, order=None, limit=None, *join):
         order = order[0] if order else [Books.sort]
         pagination = None
-        result = self.search_query(term, config, *join).order_by(*order).all()
+        
+        # search also through the subtitles (for videos)
+        searcher = lb_search.CaptionSearcher(XKLB_DB_FILE)
+        other_terms = searcher.get_search_terms(term)
+        # lb_search.get_search_terms returns a list of video titles, "term" parameter is expected to be a book/video title
+        term = [term] + other_terms
```
Comment (avni-student) [inline cps/db.py:982]: Per discussions, change variable term to a name reflective of what it's doing (something like termList or titleList ?). 

Separate search query from results/video_titles into two variables.

File: cps/db.py [hunk 3]
```diff
@@ -29,7 +29,7 @@
 from sqlite3 import OperationalError as sqliteOperationalError
 from sqlalchemy import create_engine
 from sqlalchemy import Table, Column, ForeignKey, CheckConstraint
-from sqlalchemy import String, Integer, Boolean, TIMESTAMP, Float
+from sqlalchemy import String, Integer, Boolean, TIMESTAMP, Float, or_
```
Comment (holta) [inline cps/db.py:32]: What's the trailing underscore on the end of `or_` all about — can you explain the purpose for the rest of us @deldesir ?  (Thanks!)

Comment (deldesir) [reply cps/db.py]: It was accidentally moved from a previous code and was a redundant import (already imported in https://github.com/iiab/calibre-web/blob/dcb0f14ed22ea4f78767d6cd67ddd19e8a19561e/cps/db.py#L43). It's in fact used to create logical OR expressions between multiple conditions. It allows you to combine conditions and retrieve results that satisfy any one or more of them. Example of its use is in https://github.com/iiab/calibre-web/blob/dcb0f14ed22ea4f78767d6cd67ddd19e8a19561e/cps/db.py#L955

File: cps/db.py (hunk 2 above)
Comment (deldesir) [reply cps/db.py]: term variable is from upstream. it is used throughout db.py. The PR was refactored to output ids instead of video titles.

File: cps/lb_search.py (hunk 1 above)
Comment (deldesir) [reply cps/lb_search.py:49]: Done.

**Comment by EMG70** [comment]
A new 24.04 VM was created and ran all PRs as instructed above.
SUDO IIAB-DIAGNOSTICS - https://paste.centos.org/view/b75366a2
Step 3.  ls /library/calibre-web/xklb-metadata.db  was not found. 
``` ubuntu@box:/usr/local/calibre-web-py3/library$ ls /library/calibre-web/xklb-metadata.db
ls: cannot access '/library/calibre-web/xklb-metadata.db': No such file or directory
```
meta.db is found in library directory.
ubuntu@box:/usr/local/calibre-web-py3/library$ ls
metadata.db

**Comment by holta** [comment]
@deldesir can you please make sure smoke-testing is complete?

It doesn't look like the [Test Plan](https://github.com/iiab/calibre-web/pull/244#issuecomment-2419878576) was properly smoke-tested?  According to @EMG70's https://paste.centos.org/view/b75366a2 Lines 914-923:

```
Oct 19 15:45:54 box systemd[1]: Started calibre-web.service - Calibre-Web.
Oct 19 15:45:56 box python3[26747]: Traceback (most recent call last):
Oct 19 15:45:56 box python3[26747]:   File "/usr/local/calibre-web-py3/cps.py", line 28, in <module>
Oct 19 15:45:56 box python3[26747]:     from cps.main import main
Oct 19 15:45:56 box python3[26747]:   File "/usr/local/calibre-web-py3/cps/__init__.py", line 42, in <module>
Oct 19 15:45:56 box python3[26747]:     from . import ub, db
Oct 19 15:45:56 box python3[26747]:   File "/usr/local/calibre-web-py3/cps/db.py", line 52, in <module>
Oct 19 15:45:56 box python3[26747]:     from .services.xb_utils import CaptionSearcher
Oct 19 15:45:56 box python3[26747]: ModuleNotFoundError: No module named 'cps.services.xb_utils'
Oct 19 15:45:56 box systemd[1]: calibre-web.service: Main process exited, code=exited, status=1/FAILURE
```

**Comment by deldesir** [comment]
@EMG70, I understand testing is hard. Sorry for the typos that affected your tests earlier. I smoke these PRs on my VM, my job is to understand why they don't work as expected on your VM. I read your log and I noticed the error you encountered is caused by a missing module, xb_utils. PR #256 should fix it. Here is how to do it: 
`git pull https://github.com/deldesir/calibre-web deldesir-patch-24 --no-rebase --no-edit`

Please also ensure you restart Calibre-Web after the pull by doing `systemctl restart calibre-web`. Waiting for your feedback...

**Comment by EMG70** [comment]
SUDO IIAB-DIAGNOSTICS - https://paste.centos.org/view/5a819555

**Comment by holta** [comment]
_Huge thanks @EMG70, for pushing everybody forward!_

**Comment by deldesir** [comment]
Thanks again @EMG70. It looks like PR #259 is not pulled properly. Can you double check with the command:
`git pull https://github.com/deldesir/calibre-web deldesir-xklb-schema --no-rebase --no-edit`

Also run `systemctl restart calibre-web` after.

**Comment by EMG70** [comment]
After running PR#259 
``` root@box:/usr/local/calibre-web-py3#  git pull https://github.com/deldesir/calibre-web deldesir-xklb-schema --no-rebase --no-edit
error: Pulling is not possible because you have unmerged files.
hint: Fix them up in the work tree, and then use 'git add/rm <file>'
hint: as appropriate to mark resolution and make a commit.
fatal: Exiting because of an unresolved conflict.
root@box:/usr/local/calibre-web-py3# 
```

**Comment by deldesir** [comment]
Can you do `git status` so I can know which file contains conflicts?

**Comment by EMG70** [comment]
root@box:/usr/local/calibre-web-py3# git status
On branch master
Your branch is ahead of 'origin/master' by 44 commits.
  (use "git push" to publish your local commits)

You have unmerged paths.
  (fix conflicts and run "git commit")
  (use "git merge --abort" to abort the merge)

Unmerged paths:
  (use "git add <file>..." to mark resolution)
	both modified:   cps/tasks/download.py

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	bin/
	include/
	lib/
	lib64
	pyvenv.cfg

no changes added to commit (use "git add" and/or "git commit -a")
root@box:/usr/local/calibre-web-py3#

**Comment by deldesir** [comment]
Thanks. I would like to see one more output or screenshot to pin the problem.

Open cps/tasks/download.py in a text editor doing `nano /usr/local/calibre-web-py3/cps/tasks/download.py`.
You'll see sections marked with conflict markers:

Copy paste the section where you see stuff:
<<<<<<< HEAD 
======= 
>>>>>>>

A screeshot is ok.

**Comment by EMG70** [comment]
```
 <<<<<<< HEAD
 57
 58                 # Database operations
 59                 with sqlite3.connect(XKLB_DB_FILE) as conn:
 60                     try:
 61                         # requested_file = conn.execute("SELECT path FROM media WHERE webpath = ? AND path NOT LIKE 'http%'", (self.media_url,)).fetchone()[0]
 62                         self.media_id, requested_file = conn.execute("SELECT id, path FROM media WHERE webpath = ? AND path NOT LIKE 'http%'", (self.media_url,)).fetchone()
 63
 64                         # Abort if there is not a path
 65                         if not requested_file:
 66                             log.info("No path found in the database")
 67                             error = conn.execute("SELECT error, webpath FROM media WHERE error IS NOT NULL").fetchone()
 68                             if error:
 69                                 log.error("[xklb] An error occurred while trying to download %s: %s", error[1], error[0])
 70                                 self.message = f"{error[1]} failed to download: {error[0]}"
 71                             else:
 72                                 log.error("%s failed to download: No path or error found in the database (likely the video failed due to unavailable fragments?)", self.media_url)
 73                                 self.message = f"{self.media_url_link} failed to download: No path or error found in the database (likely the video failed due to unavailable fragm>
 74                                 self.media_id = conn.execute("SELECT id FROM media WHERE webpath = ?", (self.media_url,)).fetchone()[0]
 75                                 conn.execute("DELETE FROM media WHERE webpath = ?", (self.media_url,))
 76                                 conn.execute("DELETE FROM captions WHERE media_id = ?", (self.media_id,))
 77                             return
 78                     except sqlite3.Error as db_error:
 79                         log.error("An error occurred while trying to connect to the database: %s", db_error)
 80                         self.message = f"{self.media_url_link} failed to download: {db_error}"
 81
 82                     self.message = self.message + "\n" + f"Almost done..."
 83                     response = requests.get(self.original_url, params={"requested_file": requested_file, "current_user_name": self.current_user_name, "shelf_id": self.shelf_id, "m>
 84                     if response.status_code == 200:
 85                         log.info("Successfully sent the requested file to %s", self.original_url)
 86                         file_downloaded = response.json()["file_downloaded"]
 87                         self.message = f"Successfully downloaded {self.media_url_link} to <br><br>{file_downloaded}"
 88                         new_video_path = response.json()["new_book_path"]
 89                         new_video_path = next((os.path.join(new_video_path, file) for file in os.listdir(new_video_path) if file.endswith((".webm", ".mp4"))), None)
 90                         # 2024-02-17: Dedup Design Evolving... https://github.com/iiab/calibre-web/pull/125
 91                         conn.execute("UPDATE media SET path = ? WHERE webpath = ?", (new_video_path, self.media_url))
 92                         conn.execute("UPDATE media SET webpath = ? WHERE path = ?", (f"{self.media_url}&timestamp={int(datetime.now().timestamp())}", new_video_path))
 93                         self.progress = 1.0
 94                     else:
 95                         log.error("Failed to send the requested file to %s", self.original_url)
 96                         self.message = f"{self.media_url_link} failed to download: {response.status_code} {response.reason}"
 97
 98                 conn.close()
 99
100 =======
101                 self._post_process_download()
102 >>>>>>> 95a0d554e50fef881f129b4ed2d929249dbffd17

```

**Comment by deldesir** [comment]
I think I understand what happened. I'll be back with a solution in a couple of minutes. Thanks for your patience.

**Comment by deldesir** [comment]
It shouldn't be you to do this, but I'll ask you to fix your install the below instructions. I must tell you the problem was caused by PR #255 proposing unecessary changes for download.py, not you. I understand now the order matters when testing PRs. After you fix this, I will adjust PR #255 to prevent this issue for others.

- Open the download.py if not already opened with `nano /usr/local/calibre-web-py3/cps/tasks/download.py`
- Remove ` <<<<<<< HEAD` (should be at line 56)
- Remove lines 100, 101, 102
- Save and Exit  from nano editor:
  - press Ctrl + X.
  - you will be prompted to save, press Y to confirm.
  - Press Enter to accept the current file name.
- Stage the file by doing `git add cps/tasks/download.py`
- Commit your changes by running `git commit -m "Resolved conflict in download.py"`
- Pull PR #259 again by running `git pull https://github.com/deldesir/calibre-web deldesir-xklb-schema`

**Comment by deldesir** [comment]
Hold on @EMG70

**Comment by holta** [comment]
> It shouldn't be you to do this

Indeed!  This is a struggle we're currently sorting out, and let's all be patient getting this right!  And in future:

- Testing volunteers should NEVER get involved with merge conflicts.  ✅
- Test Plan / Test Instructions need to be FULLY SMOKE-TESTED in advance.  ✅

**Comment by EMG70** [comment]
I have followed your instructions .
SUDO IIAB-DIAGNOSTICS - https://paste.centos.org/view/ba1d5609
 calibre-web still not reachable.

root@box:/usr/local/calibre-web-py3# systemctl restart calibre-web
root@box:/usr/local/calibre-web-py3# systemctl status calibre-web
× calibre-web.service - Calibre-Web
     Loaded: loaded (/etc/systemd/system/calibre-web.service; enabled; preset: enabled)
     Active: failed (Result: exit-code) since Sat 2024-10-19 22:07:10 BST; 6s ago
   Duration: 1.330s
    Process: 2247 ExecStart=/usr/local/calibre-web-py3/bin/python3 /usr/local/calibre-web-py3/cps.py -p /library/calibre-web/config/app.db (code=exited, status=1/FAILURE)
   Main PID: 2247 (code=exited, status=1/FAILURE)
        CPU: 1.328s

Oct 19 22:07:10 box python3[2247]:   File "/usr/local/calibre-web-py3/cps/__init__.py", line 42, in <module>
Oct 19 22:07:10 box python3[2247]:     from . import ub, db
Oct 19 22:07:10 box python3[2247]:   File "/usr/local/calibre-web-py3/cps/db.py", line 52, in <module>
Oct 19 22:07:10 box python3[2247]:     from .services.xb_utils import CaptionSearcher
Oct 19 22:07:10 box python3[2247]:   File "/usr/local/calibre-web-py3/cps/services/xb_utils.py", line 9, in <module>
Oct 19 22:07:10 box python3[2247]:     from cps.glue_db import GlueDB, MediaBooksMapping
Oct 19 22:07:10 box python3[2247]: ModuleNotFoundError: No module named 'cps.glue_db'
Oct 19 22:07:10 box systemd[1]: calibre-web.service: Main process exited, code=exited, status=1/FAILURE
Oct 19 22:07:10 box systemd[1]: calibre-web.service: Failed with result 'exit-code'.
Oct 19 22:07:10 box systemd[1]: calibre-web.service: Consumed 1.328s CPU time.
root@box:/usr/local/calibre-web-py3#

**Comment by deldesir** [comment]
Ok. Let's call it a failed test for now. On my side, I will be following the instructions step by step to ensure we eat the same food👌. I'll provide some feedback once done to better assist you. Thanks for your hard testing and patience.

**Comment by holta** [comment]
> following the instructions step by step to ensure we eat the same food👌

Perfect.  Thank you.

**Comment by deldesir** [comment]
Just done the test following the instructions step by step: https://paste.centos.org/view/dc30dfdf
@EMG70, please do the test again when possible.

You can either do it with a new VM or do it with the current one instead by doing
- ` cd /usr/local/`
- `mv calibre-web-py3 calibre-web-py3.OLD`
- `cd /opt/iiab/iiab`
- `./runrole --reinstall calibre-web`

... and then try the pulls again:
```
cd /usr/local/calibre-web-py3
git pull https://github.com/deldesir/calibre-web deldesir-search-captions --no-rebase --no-edit
git pull https://github.com/deldesir/calibre-web deldesir-patch-22 --no-rebase --no-edit
git pull https://github.com/deldesir/calibre-web deldesir-patch-23 --no-rebase --no-edit
git pull https://github.com/deldesir/calibre-web deldesir-patch-24 --no-rebase --no-edit
git pull https://github.com/deldesir/calibre-web deldesir-patch-25 --no-rebase --no-edit
git pull https://github.com/deldesir/calibre-web deldesir-patch-27 --no-rebase --no-edit
git pull https://github.com/deldesir/calibre-web deldesir-xklb-schema --no-rebase --no-edit
git pull https://github.com/deldesir/calibre-web deldesir-patch-31 --no-rebase --no-edit
iiab-update -f
```

Above worked for me, but it they fail for you, feel free to let me know:
image

**Comment by EMG70** [comment]
Just a quick note.I have installed another fresh VM ,this time a 24.10.It worked OK as per above original instuctions.I know @holta had suggested not to use 24.10.
Could this issue be confined to 24.10 LTS?
SUDO IIB-DIAGNOSTICS  https://paste.centos.org/view/bc3e5380

**Comment by EMG70** [comment]
> I did the test following the instructions step by step: https://paste.centos.org/view/dc30dfdf @EMG70, please do the test again when possible.
> 
> You can either do it with a new VM or do it with the current one instead by doing
> 
>     * ` cd /usr/local/`
> 
>     * `mv calibre-web-py3 calibre-web-py3.OLD`
> 
>     * `cd /opt/iiab/iiab`
> 
>     * `./runrole --reinstall calibre-web`
> 
> 
> ... and then try the pulls again:
> 
> ```
> cd /usr/local/calibre-web-py3
> git pull https://github.com/deldesir/calibre-web deldesir-search-captions --no-rebase --no-edit
> git pull https://github.com/deldesir/calibre-web deldesir-patch-22 --no-rebase --no-edit
> git pull https://github.com/deldesir/calibre-web deldesir-patch-23 --no-rebase --no-edit
> git pull https://github.com/deldesir/calibre-web deldesir-patch-24 --no-rebase --no-edit
> git pull https://github.com/deldesir/calibre-web deldesir-patch-25 --no-rebase --no-edit
> git pull https://github.com/deldesir/calibre-web deldesir-patch-27 --no-rebase --no-edit
> git pull https://github.com/deldesir/calibre-web deldesir-xklb-schema --no-rebase --no-edit
> git pull https://github.com/deldesir/calibre-web deldesir-patch-31 --no-rebase --no-edit
> iiab-update -f
> ```
> 
> Above instructions worked for me, but it they fail for you, feel free to let me know: image

Am on it now.

**Comment by deldesir** [comment]
> Just a quick note.I have installed another fresh VM ,this time a 24.10.It worked OK as per above original instuctions.I know @holta had suggested not to use 24.10. Could this issue be confined to 24.10 LTS? SUDO IIB-DIAGNOSTICS https://paste.centos.org/view/bc3e5380
> 
> Download of a single video failed though.

Glad the instructions worked for 24.10. 
Yes, you're right for the download failure per #267

**Comment by holta** [comment]
> * WARNING: Avoid Ubuntu 24.10 for now, as all video downloads appear to fail due to ["Download to IIAB" button fails every time on Ubuntu 24.10 — "An error occurred during the subprocess execution: 'NoneType' object is not subscriptable" and later "failed to download: None" — perhaps b/c xklb marked your video as deleted because it failed to download in previous tries? #267](https://github.com/iiab/calibre-web/issues/267) !

▶️ @EMG70 Please AVOID Ubunt 24.10 as explained above (yesterday):  https://github.com/iiab/calibre-web/pull/244#issuecomment-2419878576

▶️ Testing needs to begin on **Ubuntu 24.04** "so we're all eating the same food."

**Comment by EMG70** [comment]
Update after @deldesir's latest instructions:
sudo iiab-diagnostics - https://paste.centos.org/view/4f127ac4

calibre-web now working OK.
Screenshot From 2024-10-19 23-24-04

**Comment by EMG70** [comment]
> > * WARNING: Avoid Ubuntu 24.10 for now, as all video downloads appear to fail due to ["Download to IIAB" button fails every time on Ubuntu 24.10 — "An error occurred during the subprocess execution: 'NoneType' object is not subscriptable" and later "failed to download: None" — perhaps b/c xklb marked your video as deleted because it failed to download in previous tries? #267](https://github.com/iiab/calibre-web/issues/267) !
> 
> ▶️ @EMG70 Please AVOID Ubuntu 24.10 as explained above (yesterday): [#244 (comment)](https://github.com/iiab/calibre-web/pull/244#issuecomment-2419878576)
> 
> ▶️ Testing needs to begin on **Ubuntu 24.04** "so we're all eating the same food."

Noted.

**Comment by deldesir** [comment]
Instructions updated to include `pipx install datasette` in step 5.

Here are the complete instructions:

**Comment by EMG70** [comment]
I will continue steps 6-9 tomorrow.

**Comment by holta** [comment]
I just fixed the [instructions for installing datasette in the Test Plan](https://github.com/iiab/calibre-web/pull/244#issuecomment-2424260710) just above:

- `pipx install datasette` ❌ 
- `PIPX_BIN_DIR=/usr/local/bin pipx install datasette` ✅
- `PIPX_HOME=/opt/pipx PIPX_BIN_DIR=/usr/local/bin pipx install datasette` if you want to follow our [official instructions](https://github.com/iiab/calibre-web/wiki#integration-of-xklb-in-calibre-web) !

**Comment by deldesir** [comment]
@EMG70, did you get a 404 not found error when testing on 24.04. On my side, when I do the tests on Raspberry PI OS, I could not access Calibre-Web normally. I had to specify the Port 8083 to access it.

**Comment by deldesir** [comment]
Calibre-Web is now accessible on UNITTEST install on Raspberry Pi OS. All I had to do is:
```
cd /opt/iiab/iiab
./runrole network 
```

**Comment by EMG70** [comment]
> @EMG70, did you get a 404 not found error when testing on 24.04. On my side, when I do the tests on Raspberry PI OS, I could not access Calibre-Web normally. I had to specify the Port 8083 to access it.

I have not yet tried latest PR tests on Raspberry PI.

**Comment by holta** [comment]
> cd /opt/iiab/iiab
> ./runrole network

The official/recommended way of doing this is to run this as root: (the `cd` command is not necessary!)

```
iiab-network
```

(But possibly these 2 variables should be set in [/etc/iiab/local_vars.yml](https://wiki.iiab.io/go/FAQ#What_is_local_vars.yml_and_how_do_I_customize_it?) first?)

```
network_install: True
network_enabled: True
```

**Comment by deldesir** [comment]
@EMG70 Test plan [updated](https://github.com/iiab/calibre-web/pull/244#issuecomment-2424260710) to check the fix for #271
